plonk/src/lagrange.rs

```diff
@@ -0,0 +1,224 @@
+        } else {
+            let denominator = self.size_as_field_element()
+                * offset.pow([self.size() as u64 - 1])
+                * (tau - offset);
+            z_h_at_tau * denominator.inverse().unwrap()
+        }
+    }
+
+    // L_n-1(tau) = Z_H(tau) * g^-1 / (n * h^(n-1) * (tau - h * g^-1))
+    // with g^n-1 = g^-1
+    fn last_lagrange_coeff(&self, tau: F) -> F {
+        let offset = self.coset_offset();
+        let z_h_at_tau = self.evaluate_vanishing_polynomial(tau);
+        if z_h_at_tau.is_zero() {
+            if tau == self.group_gen_inv() * offset {
+                F::one()
+            } else {
+                F::zero()
+            }
+        } else {
+            let denominator = self.size_as_field_element()
+                * offset.pow([self.size() as u64 - 1])
+                * (tau - offset * self.group_gen_inv());
+            z_h_at_tau * self.group_gen_inv() * denominator.inverse().unwrap()
+        }
+    }
+
+    // similar to `EvaluationDomain::evaluate_all_lagrange_coefficients()`
+    //
+    // # Panic
+    // if `range` exceeds the `self.size()`
+    fn lagrange_coeffs_for_range(&self, range: Range<usize>, tau: F) -> Vec<F> {
+        if range.end > self.size() {
+            panic!("Out of range: domain size smaller than range.end");
+        }
+        let size = range.end - range.start;
+        let z_h_at_tau = self.evaluate_vanishing_polynomial(tau);
+        let offset = self.coset_offset();
+        let group_gen = self.group_gen();
+        let group_start = group_gen.pow([range.start as u64]);
+
+        if z_h_at_tau.is_zero() {
+            // In this case, we know that tau = hg^i, for some value i.
+            // Then i-th lagrange coefficient in this case is then simply 1,
+            // and all other lagrange coefficients are 0.
+            // Thus we find i by brute force.
+            let mut u = vec![F::zero(); size];
+            let mut omega_i = offset * group_start;
+            for u_i in u.iter_mut().take(size) {
+                if omega_i == tau {
+                    *u_i = F::one();
+                    break;
+                }
+                omega_i *= &group_gen;
+            }
+            u
+        } else {
+            // In this case we have to compute `Z_H(tau) * v_i / (tau - h g^i)`
+            // for i in start..end
+            // We actually compute this by computing (Z_H(tau) * v_i)^{-1} * (tau - h g^i)
+            // and then batch inverting to get the correct lagrange coefficients.
+            // We let `l_i = (Z_H(tau) * v_i)^-1` and `r_i = tau - h g^i`
+            // Notice that since Z_H(tau) is i-independent,
+            // and v_i = g * v_{i-1}, it follows that
+            // l_i = g^-1 * l_{i-1}
+
+            let group_gen_inv = self.group_gen_inv();
+            let start = range.start as u64;
+
+            // v_0_inv = n * h^(n-1)
+            let v_0_inv = self.size_as_field_element() * offset.pow([self.size() as u64 - 1]);
```

Comment:
Same here for `offset.pow[self.size]`.

Reply:
changed in [caaebf0](https://github.com/EspressoSystems/jellyfish/pull/611/commits/caaebf0e3f9ffb95de1df42ffa1cb97edb69fd8a)